File: mobile/test/unit/receipts_test.dart
// The same cases as web/src/lib/receipts.test.ts: the two clients must agree about what a tick
// means, or the same conversation reads differently depending on which one you opened.

import 'package:flutter_test/flutter_test.dart';
import 'package:pheme_mobile/src/chat/receipts.dart';
import 'package:pheme_mobile/src/models/chat_models.dart';
import 'package:pheme_mobile/src/models/models.dart'
    show ChannelRole, PublicUser;

const me = 'me';
const early = 1;
const mid = 2;
const late = 3;

ConversationMember member(
  String userId, {
  int delivered = 0,
  int read = 0,
}) => ConversationMember(
  id: 'm-$userId',
  conversationId: 'c1',
  userId: userId,
  role: ChannelRole.user,
  joinedAt: '2026-07-16T10:00:00Z',
  user: PublicUser(id: userId, displayName: userId),
  deliveredSeq: delivered,
  readSeq: read,
);

void main() {
  group('messageReceipt', () {
    test('is sent while the other side has not reported at all', () {
      expect(
        messageReceipt(mid, [member(me), member('bob')], me),
        Receipt.sent,
      );
    });

    test('is delivered once they have received up to it', () {
      final members = [member(me), member('bob', delivered: late)];
      expect(messageReceipt(mid, members, me), Receipt.delivered);
    });

    test('is read once they have read up to it', () {
      final members = [member(me), member('bob', delivered: late, read: late)];
      expect(messageReceipt(mid, members, me), Receipt.read);
    });

    test(
      'is still only delivered for a message newer than what they have read',
      () {
        final members = [
          member(me),
          member('bob', delivered: late, read: early),
        ];
        expect(messageReceipt(mid, members, me), Receipt.delivered);
      },
    );

    test('never waits on yourself', () {
      // We have read nothing; they have read everything. It is read.
      final members = [member(me), member('bob', delivered: late, read: late)];
      expect(messageReceipt(mid, members, me), Receipt.read);
    });

    group('groups wait for the slowest member', () {
      test('is read only once every member has read it', () {
        final members = [
          member(me),
          member('bob', delivered: late, read: late),
          member('carol', delivered: late, read: late),
        ];
        expect(messageReceipt(mid, members, me), Receipt.read);
      });

      test('is delivered, not read, while one member has only received it', () {
        final members = [
          member(me),
          member('bob', delivered: late, read: late),
          member('carol', delivered: late, read: early),
        ];
        expect(messageReceipt(mid, members, me), Receipt.delivered);
      });

      test('is sent while one member has not received it at all', () {
        final members = [
          member(me),
          member('bob', delivered: late, read: late),
          member('carol'),
        ];
        expect(messageReceipt(mid, members, me), Receipt.sent);
      });
    });

    test('claims nothing in a conversation with nobody else left in it', () {
      expect(messageReceipt(mid, [member(me)], me), Receipt.sent);
    });
  });

  group('applyReceipt', () {
    test('moves a member forward', () {
      final members = [
        member(me),
        member('bob', delivered: early, read: early),
      ];
      final next = applyReceipt(
        members,
        const ConversationReceipt(
          userId: 'bob',
          deliveredSeq: late,
          readSeq: mid,
        ),
      );
      expect(next[1].deliveredSeq, late);
      expect(next[1].readSeq, mid);
    });

    test('never moves a member backwards', () {
      final members = [member(me), member('bob', delivered: late, read: late)];
      final next = applyReceipt(
        members,
        const ConversationReceipt(
          userId: 'bob',
          deliveredSeq: early,
          readSeq: early,
        ),
      );
      expect(next[1].deliveredSeq, late);
      expect(next[1].readSeq, late);
    });

    test('leaves everyone else untouched', () {
      final members = [
        member(me),
        member('bob', delivered: early),
        member('carol', delivered: early),
      ];
      final next = applyReceipt(
        members,
        const ConversationReceipt(userId: 'bob', readSeq: late),
      );
      expect(next[2].deliveredSeq, early);
      expect(next[2].readSeq, 0);
    });
  });
}

File: web/src/lib/receipts.test.ts
import { describe, expect, it } from 'vitest'
import { applyReceipt, messageReceipt } from './receipts'
import type { ConversationMember } from './types'

const ME = 'me'
// Message sequences: strictly increasing, mirroring the hub's per-conversation counter.
const S = {
  early: 10,
  mid: 11,
  late: 12,
}

function member(userId: string, deliveredSeq?: number, readSeq?: number): ConversationMember {
  return {
    id: `m-${userId}`,
    conversationId: 'c1',
    userId,
    role: 'user',
    joinedAt: '2026-07-16T10:00:00Z',
    user: { id: userId, displayName: userId },
    deliveredSeq,
    readSeq,
  } as ConversationMember
}

describe('messageReceipt', () => {
  it('is sent while the other side has not reported at all', () => {
    const members = [member(ME), member('bob')]
    expect(messageReceipt(S.mid, members, ME)).toBe('sent')
  })

  it('is delivered once they have received up to it', () => {
    const members = [member(ME), member('bob', S.late)]
    expect(messageReceipt(S.mid, members, ME)).toBe('delivered')
  })

  it('is read once they have read up to it', () => {
    const members = [member(ME), member('bob', S.late, S.late)]
    expect(messageReceipt(S.mid, members, ME)).toBe('read')
  })

  it('is still only delivered for a message NEWER than what they have read', () => {
    const members = [member(ME), member('bob', S.late, S.early)]
    expect(messageReceipt(S.mid, members, ME)).toBe('delivered')
  })

  it('never waits on yourself — your own watermark is irrelevant', () => {
    // We have read nothing; they have read everything. It is read.
    const members = [member(ME), member('bob', S.late, S.late)]
    expect(messageReceipt(S.mid, members, ME)).toBe('read')
  })

  // The rule the user asked for: in a group, two ticks mean EVERYONE read it.
  describe('groups wait for the slowest member', () => {
    it('is read only once every member has read it', () => {
      const all = [member(ME), member('bob', S.late, S.late), member('carol', S.late, S.late)]
      expect(messageReceipt(S.mid, all, ME)).toBe('read')
    })

    it('is delivered, not read, while one member has only received it', () => {
      const members = [member(ME), member('bob', S.late, S.late), member('carol', S.late, S.early)]
      expect(messageReceipt(S.mid, members, ME)).toBe('delivered')
    })

    it('is sent while one member has not received it at all', () => {
      const members = [member(ME), member('bob', S.late, S.late), member('carol')]
      expect(messageReceipt(S.mid, members, ME)).toBe('sent')
    })
  })

  it('claims nothing in a conversation with nobody else left in it', () => {
    expect(messageReceipt(S.mid, [member(ME)], ME)).toBe('sent')
  })
})

describe('applyReceipt', () => {
  it('moves a member forward', () => {
    const members = [member(ME), member('bob', S.early, S.early)]
    const next = applyReceipt(members, { userId: 'bob', deliveredSeq: S.late, readSeq: S.mid })
    expect(next[1].deliveredSeq).toBe(S.late)
    expect(next[1].readSeq).toBe(S.mid)
  })

  // Out-of-order receipts are normal — two devices, a retry, a catch-up. An older one must not
  // un-read what is already read, exactly as the server refuses to.
  it('never moves a member backwards', () => {
    const members = [member(ME), member('bob', S.late, S.late)]
    const next = applyReceipt(members, { userId: 'bob', deliveredSeq: S.early, readSeq: S.early })
    expect(next[1].deliveredSeq).toBe(S.late)
    expect(next[1].readSeq).toBe(S.late)
  })

  it('leaves everyone else untouched', () => {
    const members = [member(ME), member('bob', S.early), member('carol', S.early)]
    const next = applyReceipt(members, { userId: 'bob', readSeq: S.late })
    expect(next[2]).toBe(members[2])
  })
})
